corehq/apps/reports/standard/deployments.py

```diff
@@ -225,15 +226,58 @@ def user_locations(self, ancestors, location_types):
             for location_type in location_types
         ]
 
+    def get_bulk_ancestors(self, location_ids):
+        """
+        Returns the grouped ancestors for the location ids passed in the
+        dictionary of following pattern
+        {location_id_1: [self, parent, parent_of_parent,.,.,.,],
+        location_id_2: [self, parent, parent_of_parent,.,.,.,],
+
+        }
+        :param domain: domain for which locations is to be pulled out
+        :param location_ids: locations ids whose ancestors needs to be find
+        :param kwargs: extra parameters
+        :return: dict
+        """
+        where = Q(domain=self.domain, location_id__in=location_ids)
+        location_ancestors = SQLLocation.objects.get_ancestors(where)
+        location_by_id = {location.location_id: location for location in location_ancestors}
+        location_by_pk = {location.id: location for location in location_ancestors}
+
+        grouped_location = {}
+        for location_id in location_ids:
+
+            location_parents = []
+            current_location = location_by_id[location_id].id if location_id in location_by_id else None
+
+            while current_location is not None:
+                location_parents.append(location_by_pk[current_location])
+                current_location = location_by_pk[current_location].parent_id
+
+            grouped_location[location_id] = location_parents
+
+        return grouped_location
+
+    def include_location_data(self):
```

Comment:
:+1:

Reply:
hey, dont you think  making this as a property rather than a normal method will be misleading or atleast would be defy the concept of a class property? Because the value it would contain is a decision value not actual data.

Reply:
Hey Cal I have resolved the request object comment.  Do you think your this comment should still be changed? Beause making it a class property and memoizing it adding very less value because its not actual data of the report and no expensive operation is being performed in it that memoization would be needed

Reply:
It is not blocking but I think it is a great fit for a calculated property (one using the `@property` decorator). It is a calculation, but it is a static one that takes no arguments returns a single value that can be seen as a property of the class, rather than a method that is doing some action. Anyway, its a minor point and mostly a person style one so I am merging but curious if @esoergel has an opinion either way on this.

Reply:
I agree with Cal's reasoning - I'd probably write this as a cached property.  It feels logically very similar to other booleans on the report class, like `exportable` or `ajax_pagination`.

@Rohit25negi I'm curious about your statement
> because its not actual data of the report

Can you elaborate on what you mean by that?  I do like the idea of a data object that's only the returned report, though I think that's unfortunately quite different from what HQ reports are now.  The report object also controls report generation.

Reply:
Ow... Yeah, by that I meant that. class attributes are supposed to contain the data/attributes of the Object which the class represents. Example: Suppose we  have two separate storage for two different projects we are maintain (couch for project x, and postgres for project y). So SQLLocation class is only suppose to have attributes like location_name, loaction_id, parent_id,location_type etc etc but should not have attributes which contains values which does not have any relation with a object of SQLLocation. for example: boolean value whether to store to this object into couch or to postgres because this decision is based on project not the location object.

Reply:
That makes sense, thanks for the explanation.  I do think unfortunately the reports framework does a poor job of separating the two concepts.  Another tack here would be to mark methods and attributes that are internal only as private variables.  That way at least they won't pollute the external-facing namespace of the class.